tests/gradients/test_finite_difference.py

```diff
@@ -0,0 +1,608 @@
+        """Test that the correct centered order 2 method is returned"""
+        coeffs, shifts = finite_diff_stencil(1, 2, "center")
+        assert np.allclose(coeffs, [-0.5, 0.5])
+        assert np.allclose(shifts, [-1, 1])
+
+    def test_correct_backward_order1(self):
+        """Test that the correct backward order 1 method is returned"""
+        coeffs, shifts = finite_diff_stencil(1, 1, "backward")
+        assert np.allclose(coeffs, [1, -1])
+        assert np.allclose(shifts, [0, -1])
+
+    def test_correct_second_derivative_forward_order1(self):
+        """Test that the correct forward order 1 method is returned"""
+        coeffs, shifts = finite_diff_stencil(2, 1, "forward")
+        assert np.allclose(coeffs, [1, -2, 1])
+        assert np.allclose(shifts, [0, 1, 2])
+
+    def test_correct_second_derivative_center_order4(self):
+        """Test that the correct forward order 4 method is returned"""
+        coeffs, shifts = finite_diff_stencil(2, 4, "center")
+        assert np.allclose(coeffs, [-2.5, 4 / 3, 4 / 3, -1 / 12, -1 / 12])
+        assert np.allclose(shifts, [0, -1, 1, -2, 2])
+
+
+class TestShiftedTapes:
+    """Tests for the generate_shifted_tapes function"""
+
+    def test_behaviour(self):
+        """Test that the function behaves as expected"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.PauliZ(0)
+            qml.RX(1.0, wires=0)
+            qml.CNOT(wires=[0, 2])
+            qml.Rot(2.0, 3.0, 4.0, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0, 2}
+        shifts = [0.1, -0.2, 1.6]
+        res = generate_shifted_tapes(tape, 1, shifts=shifts)
+
+        assert len(res) == len(shifts)
+        assert res[0].get_parameters(trainable_only=False) == [1.0, 2.0, 3.1, 4.0]
+        assert res[1].get_parameters(trainable_only=False) == [1.0, 2.0, 2.8, 4.0]
+        assert res[2].get_parameters(trainable_only=False) == [1.0, 2.0, 4.6, 4.0]
+
+
+class TestFiniteDiff:
+    """Tests for the finite difference gradient transform"""
+
+    def test_non_differentiable_error(self):
+        """Test error raised if attempting to differentiate with
+        respect to a non-differentiable argument"""
+        psi = np.array([1, 0, 1, 0]) / np.sqrt(2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.QubitStateVector(psi, wires=[0, 1])
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.probs(wires=[0, 1])
+
+        # by default all parameters are assumed to be trainable
+        with pytest.raises(
+            ValueError, match=r"Cannot differentiate with respect to parameter\(s\) {0}"
+        ):
+            finite_diff(tape)
+
+        # setting trainable parameters avoids this
+        tape.trainable_params = {1, 2}
+        dev = qml.device("default.qubit", wires=2)
+        tapes, fn = finite_diff(tape)
+        res = fn(qml.math.squeeze(dev.batch_execute(tapes)))

```

Comment:
I am still so confused about the squeeze. Is this because the device returns a result like `[1.2]` for a single expval, and a list like `[[1.2], [-0.2], [1.0]]` for a list of expvals? And the `[]` structure is there so that we potentially get `[[1.2, 0.], [-0.2, 0.9], [1.0,0.]]` if the qnode returns two values? So what the squeeze does is as follows: 
* more than one scalar returned -> produce an array/tensor of scalar results
* one scalar returned -> produce the scalar
Right?

Reply:
> I am still so confused about the squeeze

So am I. I cannot work out why it seems to be required. I think we desperately need to revisit how the device handles return types and dimensions.

I'll add a comment to this test.

Reply:
Oh, I know why... it's is because, in this particular case, `return qml.probs(wires=[0, 1])` is resulting in
```python
[[0.5432, 0.123, 0.342, 0.5346]]
```
Without the squeeze, this results in a Jacobian of shape `(4, 1, 2)`. With the squeeze, the Jacobian is of size `(4, 2)`.

Essentially, this treats the _elements_ of the probability vector the same as the floats returned from `expval`.

Having said that, I don't like this behaviour, and still think we should revisit! Ideally, I would prefer the following:

```python
...     return qml.expval(qml.PauliZ(0)), qml.probs(wires=[0, 1])
>>> qnode(params)
(0.42123, tensor([0.1, 0.2, 0.1, 0.6])
```

Reply:
Agree!